### ISSUE TITLE
In certain scenario HttpRequestDecoder does not decode messages the same way other HTTP Servers do

### ISSUE THREAD

**Issue by blucas** [netty/netty#1815]
Hi,

I happened to stumble upon a certain scenario whereby Netty's HttpRequestDecoder (Netty v3.6.2) acts differently to other HTTP Servers (such as nginx).

Given the content below has been sent to a Nginx server

```
GET / HTTP/1.0\r\n
Host: 192.168.0.237\r\n
User-Agent: ApacheBench/2.3\r\n
Accept: */*\r\n
\r\n
GET / HTTP/1.0\r\n
Host: 192.168.0.237\r\n
User-Agent: ApacheBench/2.3\r\n
Accept: */*\r\n
\r\n
GET / HTTP/1.0\r\n
Host: 192.168.0.237\r\n
User-Agent: ApacheBench/2.3\r\n
Accept: */*\r\n
\r\n
```

Nginx will respond with a single HttpResponse.  I believe this is because Nginx reads the first request, knows it does not have a body (no Content-Length header), and thus ignores the rest of the content in it's buffer. Then sends the first request down its pipeline for processing. 

Given the same content, Netty acts differently.
It reads the content into a `ChannelBuffer` (see `NioWorker.read(SelectionKey k)`)

`HttpRequestDecoder` will then generate 3 `HttpRequest`s and call `fireMessageReceived` for each.  Netty will then respond with 3 separate `HttpResponse`s.

The reason Netty generates 3 `HttpRequest`s is because of the `while (input.readable())` loop in `ReplayingDecoder` (see below).  `input` contains the content (mentioned above) sent from a client.  So even though a `result` has been generated, the loop will continue because there is more data in `input` that needs to be processed. 

``` java
private void callDecode(
   ChannelHandlerContext context, Channel channel,
   ChannelBuffer input, ChannelBuffer replayableInput, SocketAddress remoteAddress) throws Exception {
   while (input.readable()) {
      int oldReaderIndex = checkpoint = input.readerIndex();
      Object result = null;
      T oldState = state;
      try {
         result = decode(context, channel, replayableInput, state);
         if (result == null) {
            if (oldReaderIndex == input.readerIndex() && oldState == state) {
               throw new IllegalStateException(
                  "null cannot be returned if no data is consumed and state didn't change.");
            } else {
               // Previous data has been discarded or caused state transition.
               // Probably it is reading on.
               continue;
            }
         }
      } catch (ReplayError replay) {
         // Return to the checkpoint (or oldPosition) and retry.
         int checkpoint = this.checkpoint;
         if (checkpoint >= 0) {
            input.readerIndex(checkpoint);
         } else {
            // Called by cleanup() - no need to maintain the readerIndex
            // anymore because the buffer has been released already.
         }
      }

      if (result == null) {
         // Seems like more data is required.
         // Let us wait for the next notification.
         break;
      }

      if (oldReaderIndex == input.readerIndex() && oldState == state) {
         throw new IllegalStateException(
            "decode() method must consume at least one byte " +
               "if it returned a decoded message (caused by: " +
               getClass() + ')');
      }

      // A successful decode
      unfoldAndFireMessageReceived(context, remoteAddress, result);
   }
}
```

I believe Netty should act like Nginx (IIS does the same thing, and I'm guessing so does Apache). For Nginx I was unfortunately not able to find the source code that handles processing the raw bytes from the socket and transforming them into a HttpRequest.

Basically I think it is invalid for a client to send multiple HTTP Requests in a single write operation. I believe Netty should handle this (like Nginx does) by silently dropping the 'extra' requests.

If you'd like to reproduce the issue:
1. Install ApacheBench (this client will generate the content I mentioned previously)
2. Run the HTTP Snoop Server Example
3. Execute the following ApacheBench command: `ab -n 1 -c 1 http://127.0.0.1:8080/`
4. Try the same command on an Nginx server and compare the results.
